java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -856,6 +859,48 @@ public void visitVariable(VariableTree tree) {
     }
   }
 
+  @Override
+  public void visitForEachStatement(ForEachStatement tree) {
+    scan(tree.variable());
+    scan(tree.expression());
+    TypeTree typeTree = tree.variable().type();
+    JavaType iterableType = (JavaType) tree.expression().symbolType();
+    if (typeTree.is(Tree.Kind.VAR_TYPE)) {
+      JavaType iteredObjectType = getIteredObjectType(iterableType);
+      setInferedType(upwardProjection(iteredObjectType), (DeferredType) typeTree.symbolType());
+    }
+    // scan the body only after handling type of variable
+    scan(tree.statement());
+  }
+
+  private JavaType getIteredObjectType(JavaType type) {
+    return getIteredObjectType(type, new HashSet<>());
+  }
+
+  private JavaType getIteredObjectType(JavaType type, Set<String> knownTypes) {
```

Comment:
Why a set of string and not a set of types ?

Reply:
Because I will have the same fully qualified name for parametrized types and raw type. I prefer to explore them only once, even if multiple parametrization exists.